### ISSUE TITLE
Goal Tracker: Mark/filter tasks as "Blocked" (Feature Request)

### ISSUE THREAD

**Issue by Marstead** [slaytostay/region-locker#17]
Right now, there are only 2 task states -- incomplete or complete. However, when doing a Chunk Locked account, there are often also interesting BLOCKED tasks, which cannot be completed until more content is unlocked. It would be nice to be able to flag added tasks as Blocked (perhaps showing in Yellow text) and to have a checkmark option in the Plugin options to hide blocked tasks.

**Comment by aHooder** [comment]
Maybe if you could find an icon for this, we could add a third state to the checkbox to mark a task as blocked, perhaps by holding shift.

**Comment by Marstead** [comment]
Would a simple 🚫 Prohibited Emoji work?

**Comment by aHooder** [comment]
Probably yeah, but ideally it should match the dimensions of the other icons in the resource folder.

**Comment by Marstead** [comment]
I can mock one up, can you post any of those icons here so I can match the dimensions?

**Comment by aHooder** [comment]
> I can mock one up, can you post any of those icons here so I can match the dimensions?

The checkbox icons are 18x18 and can be found in the repository here: [unchecked](https://github.com/slaytostay/region-locker/blob/master/src/main/resources/com/goaltracker/checkbox_icon.png), [checked](https://github.com/slaytostay/region-locker/blob/master/src/main/resources/com/goaltracker/checked_icon.png)

**Comment by aHooder** [comment]
They probably originate from RuneLIte, so maybe they have an appropriate icon for this as well.

**Comment by aHooder** [comment]
This one perhaps?
![nope_icon](https://user-images.githubusercontent.com/831317/226135180-64f892f0-9021-4b9d-b4d5-3a318e11e854.png)

**Comment by aHooder** [comment]
Some options:
A:
![image](https://user-images.githubusercontent.com/831317/226138624-814517fc-b584-493f-9a3e-4185db5e0ba3.png)
B:
![image](https://user-images.githubusercontent.com/831317/226138569-89b90ec4-d2dc-4d12-9e48-daec95482119.png)
C:
![image](https://user-images.githubusercontent.com/831317/226138648-3164fc15-b9af-4dd7-834f-ffbff687d5bf.png)
D:
![image](https://user-images.githubusercontent.com/831317/226138550-bce91be8-f37f-478f-a7d2-27b8d50420e3.png)

**Comment by aHooder** [comment]
And yet another alternative, though it's probably more applicable to something like a partially completed goal:
![image](https://user-images.githubusercontent.com/831317/226139791-fdedb1f3-6609-4195-b15c-5edf68b3ad4c.png)

**Comment by Marstead** [comment]
Hey there; just saw these. I think the 🚫 icon looks great and clear that it would be marking blocked tasks. I guess just to be clear from the mockup, this would be marking an entire Goal **category** as Complete, In Progress, or Blocked, right? Would it also be possible to assign Complete & Blocked to individual **requirements** as well?

The way I'd envision it working:

1) If all requirements are marked complete, the category box is marked with a checkbox
2) If NO requirements are marked complete, the category box is empty
3) If at least one requirement is marked complete but not all requirements are marked complete, the category box is marked with a "-" (as in your mock-up, "In-Progress")
4) If at least one requirement is marked blocked, the category box is marked blocked (🚫)

As far as how you'd mark individual requirements to one of these 3 states (Not started, complete, blocked) I envision they would start as "Not Started" (yellow), click once to "Complete" (green), click again to "Blocked" (red), click again to wrap back around to "Not Started" (yellow). 

That'd be a change from the current behavior where they start as "Not Started" (red), click once to "Complete" (green), click again to wrap back to "Not Started" (red).

**Comment by Marstead** [comment]
Thanks for your work on this by the way especially so early, this is a big-time "nice to have" but not urgent in any way.

**Comment by Marstead** [comment]
To describe the use case, for a Chunk-locked account, say you add 10 requirements to a chunk to mark it completed. It is only completed when all 10 tasks are done. If you haven't done anything it's not started (empty checkbox). If you complete just one task out of 10 it's in progress. If you have even one blocked task, regardless of all other task statuses, the entire chunk is blocked from being completed.

There are some differences between different chunk accounts' rules, for some accounts they simply ignore blocked tasks until they become unblocked. For example, if a chunk contains a Runecrafting altar but it is impossible for that account to do Runecrafting yet due to insufficient Chunk unlocks, that account might mark the altar as a "Blocked Requirement" but then continue to do the unblocked tasks in the Chunk until they are all complete. In this case they might want the Chunk to show as "Completed" when all requirements except blocked requirements have been completed. 

Maybe that can be a checkbox setting--"Blocked Requirements stop Chunk from Completing".

Let me know if that all made sense or any clarification is needed. Thanks again!

**Comment by aHooder** [comment]
That all sounds pretty sensible, yeah. Currently I have only implemented goal blocking per top level goal, not per requirement, but it's probably simple enough to extend. I think unless it's too confusing, it could make sense to mark stuff as blocked by shift clicking though.

I'm also thinking it could be useful to allow modifying the overall goal state directly, in addition to having it automatically update as sub tasks/requirements are completed. Then people are free to not need sub goals, or to override the automatic state tracking. Currently there is no auto updating of the overall task state when requirements are completed, so that'll have to be implemented as well.

If I remember correctly, requirements are currently color coded similar to quests in-game, with red being not completed, and green being completed. Maybe it would be logical to add a yellow middle state to mean "in progress", and possibly another color entirely for blocked requirements? Or maybe just prefixing it with a blocked icon for instance?

I'm open to doing it how the chunk community wants it to be done though, I'm only throwing my thoughts into the mix. If you'd want to ask around a bit to hear what people think, that might be a good idea.

I'll probably wait to submit these changes until some time after the currently pending plugin hub update gets accepted, so we're not really in a rush. I should also mention that I'm not really looking to delve super deeply into expanding the region locker plugin for the time being, but I figured I would at least look through the issue tracker and clear up the main pain points.

**Comment by Marstead** [comment]
Sounds good--I will reach out in the Picker V2 discord and see what folks think.

**Comment by Marstead** [comment]
Brought it up in the Discord and got some discussion + reacts, here is what I think makes sense:

**REQUIREMENTS (Tasks)**
Currently, requirements have only two states -- red & green for incomplete and complete.
Change this to three states -- red (blocked), yellow (not complete/in progress), and green (complete).
No changes to logic needed, the player would simply click cycle through yellow (default) to green (complete) to red (blocked).

**CHUNK STATUS**
Currently, the chunk status is a checkbox that is either empty or checked.
Change this to three states -- 🚫 (blocked), ➖ (not complete/in progress), ✅ (complete)
Again, no changes to logic needed (doesn't need to pull from requirements). The player would simply click cycle through ➖ (default), ✅ (complete), 🚫 (blocked)

**MAP DRAWING**
There's currently a plugin setting to draw a colored box around incomplete chunks on the map. It might be nice to be able to draw a box for both incomplete and blocked chunks separately.

Thanks!

**Comment by aHooder** [comment]
I've implemented this in 278cc81996aa5977ee5a9cc2dd7ddeb90b31f003 finally. If you wouldn't mind testing things for a bit before we submit it for another release, I'd be happy to help with setting up a testing environment.

**Comment by aHooder** [comment]
Here's what it ended up looking like btw: 
![image](https://user-images.githubusercontent.com/831317/229337681-014f36f6-4827-4f2f-9ecf-3837ba417eca.png)

**Comment by Marstead** [comment]
That looks amazing! I'm happy to help test, just hit me up here or on Discord @ Marstead#0001.
